dynamiqs/utils/utils.py

```diff
@@ -620,6 +626,30 @@ def dm_fidelity(x: Tensor, y: Tensor) -> Tensor:
     return trace_sqrtm_tmp.pow(2).real
 
 
+def _ket_dm_fidelity(x: Tensor, y: Tensor) -> Tensor:
+    r"""Returns the fidelity between a pure state and a density matrix
+
+    The fidelity of a pure state $\rho$ and $\psi$ is defined by:
+
+    $$
+        F(\rho, \sigma) = \tr{\sqrt{\sqrt{\rho}\sigma\sqrt{\rho}}}^2.
+    $$
+    where $\sigma = \ket{\psi}\bra{\psi}$  and simplifiable to
+    $
+        F(\rho, \sigma) = \tr{\bra{\psi}\rho\ket{\psi}}^2.
+    $$
+
+    Args:
+        x _(..., n, n)_: A density matrix.
+        y _(..., n, n)_: A ket.
+
+    Returns:
+        _(...)_ Real-valued fidelity.
+
+    """
+    return trace(ket_to_bra(x) @ y @ x).real
```

Comment:
```suggestion
    return (ket_to_bra(x) @ y @ x).squeeze(-1).squeeze(-1).real
```
Does the same, but is more clear on what we want to do on the tensor IMO (the resulting shape of `ket_to_bra(x) @ y @ x` is `(..., 1, 1)` which we want to squeeze. I'm scared the trace is a bit confusing becasue it appears in the definition of fidelities, but we don't actually compute the trace of any matrix here, it's just a hidden way to squeeze the result.

Reply:
You can also `.squeeze(-2, -1)` in a single call.

Also, if we take the real part, we might need a check that `y` is hermitian, and to return a warning to say we take the real part if it isn't (in case people want to call that for operators).

Reply:
> You can also .squeeze(-2, -1) in a single call.

Only Pytorch 2.0 unfortunately.

Reply:
I think I disagree. You wrote that `_ket_dm_fidelity` computes `<x|y|x>` which is by definition `Tr(y |xXx|)` so it makes perfect sense to use `trace` here

Reply:
If you actually compute `<x|y|x>` then I agree we should not take the trace. Even if it's the same, since `Tr[<x|y|x>] = <x|y|x>`. And you're not computing `Tr[y |x><x|]` here which would equivalent but slower.